packages/@aws-cdk/cdk/lib/reference.ts

```diff
@@ -1,116 +1,26 @@
-import { ResolveContext, Token } from "./token";
+import { Token } from "./token";
 
 /**
- * A Token that represents a CloudFormation reference to another resource
+ * A Token that represents a reference between two constructs
  *
- * If these references are used in a different stack from where they are
- * defined, appropriate CloudFormation `Export`s and `Fn::ImportValue`s will be
- * synthesized automatically instead of the regular CloudFormation references.
- *
- * Additionally, the dependency between the stacks will be recorded, and the toolkit
- * will make sure to deploy producing stack before the consuming stack.
- *
- * This magic happens in the prepare() phase, where consuming stacks will call
- * `consumeFromStack` on these Tokens and if they happen to be exported by a different
- * Stack, we'll register the dependency.
+ * References are recorded.
  */
 export class Reference extends Token {
   /**
    * Check whether this is actually a Reference
    */
   public static isReferenceToken(x: Token): x is Reference {
-    return (x as any).consumeFromStack !== undefined;
+    return (x as any).isReference === true;
   }
 
-  public readonly isReference?: boolean;
-
-  /**
-   * What stack this Token is pointing to
-   */
-  private readonly producingStack?: Stack;
+  public readonly isReference: boolean = true;
 
-  /**
-   * The Tokens that should be returned for each consuming stack (as decided by the producing Stack)
-   */
-  private readonly replacementTokens: Map<Stack, Token>;
+  public readonly target: Construct;
 
-  constructor(value: any, displayName?: string, scope?: Construct) {
-    if (typeof(value) === 'function') {
-        throw new Error('Reference can only hold CloudFormation intrinsics (not a function)');
-    }
-    // prepend scope path to display name
-    if (displayName && scope) {
-      displayName = `${scope.node.path}.${displayName}`;
-    }
+  constructor(value: any, displayName: string, target: Construct) {
```

Comment:
I still feel we could automatically let `Reference` tokens know that they are being used instead of explicit `consumeFromStack`

Reply:
To me, that's the essence of a reference token. Then `CfnReference` utilizes this capability to do the cross-stack magic, but the behavior is polymorphic. Otherwise, what's the added value in actually modeling the concept of a reference?